### ISSUE TITLE
[Feature]: Add grouping or division for a test names within the same 'describe' section in the console output

### ISSUE THREAD

**Issue by AlexKomanov** [microsoft/playwright#7601]
> ### Feature request

Currently, if you run with two tests within the same 'describe' section - the names of the 'describe' and test are written in one line in the console output without any grouping/division.
Example of the current state (see also the added screenshots):
describe name  test1 name
describe name test2 name
![image](https://user-images.githubusercontent.com/52656401/125572024-54983e41-edba-472e-aa75-3ee62105c755.png)
![image](https://user-images.githubusercontent.com/52656401/125572195-38098440-dd86-4b52-a4fb-c1f33e304dc8.png)
![image](https://user-images.githubusercontent.com/52656401/125572225-fb758870-4ddc-4876-900f-6ecfaee35810.png)
![image](https://user-images.githubusercontent.com/52656401/125572278-93e29df8-e31c-4fd8-850c-2a55ade5c5ff.png)

I suggest any grouping of the tests within the same 'describe' part. For example:
describe name
      -  test1 name
      -  test2 name
I use the 1.12.3 Playwright version with a Windows platform. My IDE is a Visual Studio Code

**Comment by aslushnikov** [comment]
@AlexKomanov when running tests in parallel, tests from different suites might complete. Having a group-structure in the console won't make it easy for the users to follow what's going on where - thus the current style of reporting.

**Comment by AlexKomanov** [comment]
@aslushnikov 
Something like in the example from here:
[https://playwright.tech/blog/tracking-frontend-coverage-with-playwright](https://playwright.tech/blog/tracking-frontend-coverage-with-playwright)

**Comment by pavelfeldman** [comment]
We are about to publish the new test reporters API in v1.13, you can experiment in the rendering there and offer a reporter for us in the PR!

**Comment by AlexKomanov** [comment]
@pavelfeldman 
Thanks.
Do you know anything about the date of release of v1.13?

**Comment by pavelfeldman** [comment]
Should be today-tomorrow.

**Comment by klhex** [comment]
@aslushnikov @pavelfeldman Mocha has a really nice standard reporting layout which is especially useful for e2e tests whose execution log is read and used by non-technical people (in our case functional testers). Also, if Mocha's parallel execution mode is activated, the log output stays orderly and structured (maybe Mocha does parallelization differently than PW Test).

To give you a visual example:

<details> <summary>PW Test Code</summary>

```javascript
// playwright-test stuff.
import { test, expect } from '@playwright/test';

test.describe('Test Case 123: Apply for consumer loan', () => {
   test.describe('Page 1 - Personal Data', () => {
    test('Correct page is shown', async () => {
    });
    test('Existing customer question is answered', async () => {
    });
    test('Mandatory fields are filled with valid data', async () => {
    });
    test('Navigation to next page', async () => {
    });
  });

  test.describe('Page 2 - Contact Data', () => {
    test('Correct page is shown', async () => {
    });
    test('Mandatory fields are filled with valid data', async () => {
    });
    test('Navigation to next page', async () => {
    });
  });

  test.describe('Page 3 - Profession & Financial Data', () => {
    test('Correct page is shown', async () => {
    });
    test('Proof of income is uploaded', async () => {
    });
    test('Mandatory fields are filled with valid data', async () => {
    });
    test('All document links can be opened', async () => {
    });
    test('Navigation to next page', async () => {
    });
  });
});
```
</details>

<details> <summary>PW Test Console Log</summary>

```bash
❯ npx playwright test src/e2e/pw-test-layout.spec.ts

Running 12 tests using 1 worker

  ✓ src/e2e/pw-test-mocha.spec.ts:8:5 › [chromium] Test Case 123: Apply for consumer loan Page 1 - Personal Data Correct page is shown (2ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:11:5 › [chromium] Test Case 123: Apply for consumer loan Page 1 - Personal Data Existing customer question is answered (0ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:14:5 › [chromium] Test Case 123: Apply for consumer loan Page 1 - Personal Data Mandatory fields are filled with valid data (0ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:17:5 › [chromium] Test Case 123: Apply for consumer loan Page 1 - Personal Data Navigation to next page (1ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:22:5 › [chromium] Test Case 123: Apply for consumer loan Page 2 - Contact Data Correct page is shown (0ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:25:5 › [chromium] Test Case 123: Apply for consumer loan Page 2 - Contact Data Mandatory fields are filled with valid data (0ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:28:5 › [chromium] Test Case 123: Apply for consumer loan Page 2 - Contact Data Navigation to next page (0ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:33:5 › [chromium] Test Case 123: Apply for consumer loan Page 3 - Profession & Financial Data Correct page is shown (0ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:36:5 › [chromium] Test Case 123: Apply for consumer loan Page 3 - Profession & Financial Data Proof of income is uploaded (1ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:39:5 › [chromium] Test Case 123: Apply for consumer loan Page 3 - Profession & Financial Data Mandatory fields are filled with valid data (0ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:42:5 › [chromium] Test Case 123: Apply for consumer loan Page 3 - Profession & Financial Data All document links can be opened (0ms)
  ✓ src/e2e/pw-test-mocha.spec.ts:45:5 › [chromium] Test Case 123: Apply for consumer loan Page 3 - Profession & Financial Data Navigation to next page (0ms)
```
</details>

As you can see above, the PW Test console log is not so easy to discern, has quite some redundancy, and could be more user-friendly esp. for non-tech people.

Now let's have a look at how Mocha does it:

<details> <summary>Mocha Test Code</summary>

```javascript
describe('Test Case 123: Apply for consumer loan', () => {
   describe('Page 1 - Personal Data', () => {
    it('Correct page is shown', async () => {
    });
    it('Existing customer question is answered', async () => {
    });
    it('Mandatory fields are filled with valid data', async () => {
    });
    it('Navigation to next page', async () => {
    });
  });

  describe('Page 2 - Contact Data', () => {
    it('Correct page is shown', async () => {
    });
    it('Mandatory fields are filled with valid data', async () => {
    });
    it('Navigation to next page', async () => {
    });
  });

  describe('Page 3 - Profession & Financial Data', () => {
    it('Correct page is shown', async () => {
    });
    it('Proof of income is uploaded', async () => {
    });
    it('Mandatory fields are filled with valid data', async () => {
    });
    it('All document links can be opened', async () => {
    });
    it('Navigation to next page', async () => {
    });
  });
});
```
</details>

<details> <summary>Mocha Console Log (w/o parallelization)</summary>

```bash
❯ npx mocha -p --require ts-node/register src/e2e/mocha-layout.ts

  Test Case 123: Apply for consumer loan
    Page 1 - Personal Data
      ✔ Correct page is shown
      ✔ Existing customer question is answered
      ✔ Mandatory fields are filled with valid data
      ✔ Navigation to next page
    Page 2 - Contact Data
      ✔ Correct page is shown
      ✔ Mandatory fields are filled with valid data
      ✔ Navigation to next page
    Page 3 - Profession & Financial Data
      ✔ Correct page is shown
      ✔ Proof of income is uploaded
      ✔ Mandatory fields are filled with valid data
      ✔ All document links can be opened
      ✔ Navigation to next page

  12 passing (2s)
```
</details>

Isn't that beautiful? 😄 

Now I am using Mocha's parallel execution mode in order to run 3 copies of the same content (3 files with same content) in parallel:

<details> <summary>Mocha Console Log (with parallelization)</summary>

```bash
❯ npx mocha --parallel --require ts-node/register src/e2e/mocha-layout*

  Test Case 123: Apply for consumer loan
    Page 1 - Personal Data
      ✔ Correct page is shown
      ✔ Existing customer question is answered
      ✔ Mandatory fields are filled with valid data
      ✔ Navigation to next page
    Page 2 - Contact Data
      ✔ Correct page is shown
      ✔ Mandatory fields are filled with valid data
      ✔ Navigation to next page
    Page 3 - Profession & Financial Data
      ✔ Correct page is shown
      ✔ Proof of income is uploaded
      ✔ Mandatory fields are filled with valid data
      ✔ All document links can be opened
      ✔ Navigation to next page


  Test Case 123: Apply for consumer loan
    Page 1 - Personal Data
      ✔ Correct page is shown
      ✔ Existing customer question is answered
      ✔ Mandatory fields are filled with valid data
      ✔ Navigation to next page
    Page 2 - Contact Data
      ✔ Correct page is shown
      ✔ Mandatory fields are filled with valid data
      ✔ Navigation to next page
    Page 3 - Profession & Financial Data
      ✔ Correct page is shown
      ✔ Proof of income is uploaded
      ✔ Mandatory fields are filled with valid data
      ✔ All document links can be opened
      ✔ Navigation to next page


  Test Case 123: Apply for consumer loan
    Page 1 - Personal Data
      ✔ Correct page is shown
      ✔ Existing customer question is answered
      ✔ Mandatory fields are filled with valid data
      ✔ Navigation to next page
    Page 2 - Contact Data
      ✔ Correct page is shown
      ✔ Mandatory fields are filled with valid data
      ✔ Navigation to next page
    Page 3 - Profession & Financial Data
      ✔ Correct page is shown
      ✔ Proof of income is uploaded
      ✔ Mandatory fields are filled with valid data
      ✔ All document links can be opened
      ✔ Navigation to next page

  36 passing (2s)
```
</details>

Even if PW Test's parallelization architecture would render orderly / sequential display of the execution log impossible, I think having the option of such a nested layout at least for non-parallelized tests would be a big plus.

If the upcoming reporter API makes creating a corresponding custom reporter layout possible -- great! (I hope you don't need to be a coder whiz in order to achieve this because otherwise I probably wouldn't be able to do this on my own.)

Either way, I really would love to see such a nested layout (name suggestion: "nlist" (for "nested list") become part of PW Test sooner rather than later.

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

We are prioritizing the features based on the upvotes, recency and activity in the issue thread. It looks like this issue only has a handful of upvotes, has not been touched recently and/or we lack sufficient feedback to act on it. We are closing issues like this one to keep our bug database maintainable. Please feel free to open a new issue and link this one to it if you think this is a mistake.
